File: javascript/packages/linter/src/rules/a11y-svg-has-accessible-text.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"

import { hasAttribute, getStaticAttributeValue, getTagLocalName, isHTMLElementNode } from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLElementNode, ParseResult, ParserOptions } from "@herb-tools/core"

class SvgHasAccessibleTextVisitor extends BaseRuleVisitor {
  visitHTMLElementNode(node: HTMLElementNode): void {
    this.checkSVGElement(node)
    super.visitHTMLElementNode(node)
  }

  private checkSVGElement(node: HTMLElementNode): void {
    const tagName = getTagLocalName(node)

    if (tagName !== "svg") return
    if (this.hasAriaHidden(node)) return

    if (hasAttribute(node, "aria-label")) return
    if (hasAttribute(node, "aria-labelledby")) return
    if (this.hasDirectTitleChild(node)) return

    this.addOffense(
      "`<svg>` must have accessible text. Set `aria-label`, or `aria-labelledby`, or nest a `<title>` element. If the `<svg>` is decorative, hide it with `aria-hidden=\"true\"`.",
      node.tag_name!.location,
    )
  }

  private hasAriaHidden(node: HTMLElementNode): boolean {
    if (!hasAttribute(node, "aria-hidden")) return false

    const value = getStaticAttributeValue(node, "aria-hidden")
    if (value === null) return true

    return value === "true"
  }

  private hasDirectTitleChild(node: HTMLElementNode): boolean {
    const children = node.body ?? []

    return children.some(child => isHTMLElementNode(child) && getTagLocalName(child) === "title")
  }
}

export class A11ySVGHasAccessibleTextRule extends ParserRule {
  static ruleName = "a11y-svg-has-accessible-text"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: false,
      severity: "warning"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new SvgHasAccessibleTextVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/a11y_svg_has_accessible_text.rs
use crate::utils::tag_utils::{element_has_aria_hidden, element_tag_name_location, get_tag_local_name, has_element_attribute};

use herb::nodes::{AnyNode, HTMLElementNode};
use herb::Visitor;

rule_visitor!(SvgHasAccessibleTextVisitor);
define_parser_rule!(
  A11ySVGHasAccessibleTextRule,
  "a11y-svg-has-accessible-text",
  Warning,
  SvgHasAccessibleTextVisitor,
  enabled: false,
  parser_options: { action_view_helpers: true }
);

impl SvgHasAccessibleTextVisitor {
  fn has_direct_title_child(node: &HTMLElementNode) -> bool {
    node.body.iter().any(|child| match child {
      AnyNode::HTMLElementNode(element) => get_tag_local_name(element).as_deref() == Some("title"),
      _ => false,
    })
  }
}

impl Visitor for SvgHasAccessibleTextVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    if get_tag_local_name(node).as_deref() == Some("svg")
      && !element_has_aria_hidden(node)
      && !has_element_attribute(node, "aria-label")
      && !has_element_attribute(node, "aria-labelledby")
      && !Self::has_direct_title_child(node)
    {
      self.add_offense(
        "`<svg>` must have accessible text. Set `aria-label`, or `aria-labelledby`, or nest a `<title>` element. If the `<svg>` is decorative, hide it with `aria-hidden=\"true\"`.",
        element_tag_name_location(node),
      );
    }

    self.walk_html_element_node(node);
  }
}
